File: IPWorks OpenPGP Samples/OpenPGP/java/openpgp.java
/*
 * IPWorks OpenPGP 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks OpenPGP in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksopenpgp
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import ipworksopenpgp.*;

public class openpgp extends ConsoleDemo {
  Openpgp pgp = new Openpgp();
  Keymgr keymgr = new Keymgr();

  public openpgp(){
    try
    {
      String keyringDir = "";
    	
      System.out.println("************************************************************");
      System.out.println("IPWorks OpenPGP Java Edition - OpenPGP Demo Application.  ");
      System.out.println("This demo shows how to use the OpenPGP component encrypt or ");
      System.out.println("decrpyt a message and to sign or verify a message.          ");
      System.out.println("************************************************************\n");

      keymgr.addKeymgrEventListener(new DefaultKeymgrEventListener() {
        public void keyList(KeymgrKeyListEvent e) {
          System.out.println(String.format("%-80s [%-8s] %-12s", e.userId, e.keyId, (e.hasSecretKey ? "private" : "")));
        }
      });

      
      boolean keyringLoaded = false;
      while(!keyringLoaded) {
    	try {
          keyringDir = prompt("Keyring Directory (Enter \":q\" to exit)");
          if(keyringDir.equalsIgnoreCase(":q"))
        	System.exit(0);
          keymgr.loadKeyring(keyringDir);
          keyringLoaded = true;
    	} catch(Exception ex) {
        	System.out.println("Error: " + ex.getMessage());      
          }
      }
      
      if (ask("Create a new key", "?", "(y/n)") == 'y') {
        System.out.println("\nListing keys...\n");
        System.out.println(String.format("%-80s %-10s %-12s", "User ID", "Key ID", "Private Key"));
        keymgr.listKeys();
        System.out.println();

        while (true) {
          char command =  ask("Please enter a command", ":", "(c: create key, d: delete key, s: save keyring, q: cancel)");
          if (command == 'c') {
            keymgr.createKey(prompt("User ID"), prompt("Passphrase"));
            System.out.println();
            keymgr.listKeys();
            System.out.println();
          } else if (command == 'd') {
            keymgr.deleteKey(prompt("User ID"));
            System.out.println();
            keymgr.listKeys();
            System.out.println();
          } else if (command == 's') {
            keymgr.saveKeyring(keyringDir);
            break;
          } else if (command == 'q') {
            break;
          } else {
            System.out.println("Invalid command.");
          }
        }
      }

      System.out.println("\nListing keys...\n");
      System.out.println(String.format("%-80s %-10s %-12s", "User ID", "Key ID", "Private Key"));
      keymgr.listKeys();
      System.out.println();

      String privateKey = prompt("Private Key User ID (Used to sign outgoing messages and decrypt incoming messages)");
      String passphrase = "";
      if (privateKey.length() > 0) passphrase = prompt("Passphrase");
      String recipientKey = prompt("Recipient Key User ID (Used to encrypt outgoing messages)");
      String signerKey = prompt("Signer Key User ID (Used to verify the signature of an incoming message)");

      while (true) {
        System.out.println();
        int command = Character.getNumericValue(ask("Please enter a command", ":", "(0: quit, 1: encrypt, 2: decrypt, 3: sign, 4: verify, 5: sign and encrypt, 6: decrypt and verify)"));
        if (command < 0 || command > 6) {
          System.out.println("Invalid Command.");
          continue;
        }
        if (command == 0) {
          break;
        }
        try {
          pgp.reset();
          pgp.setASCIIArmor(true);
          if (ask("Read from file or string", "?", "(f/s)") == 'f') {
            pgp.setInputFile(prompt("Input File", ":", pgp.getInputFile()));
            pgp.setOutputFile(prompt("Output File", ":", pgp.getOutputFile()));
            pgp.setOverwrite(ask("Overwrite file", "?", "(y/n)") == 'y');
          } else {
            System.out.println("Please enter the message. When finished enter \":q\" on a line by itself:");
            BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
            String temp = "";
            while (!(temp = bf.readLine()).equalsIgnoreCase(":q")) {
              temp += "\n";
              byte[] bytes = new byte[pgp.getInputMessage().length + temp.getBytes().length];
              System.arraycopy(pgp.getInputMessage(), 0, bytes, 0, pgp.getInputMessage().length);
              System.arraycopy(temp.getBytes(), 0, bytes, pgp.getInputMessage().length, temp.getBytes().length);
              pgp.setInputMessage(bytes);
            }
          }

          if (command == 1) { //Encrypt
            if (recipientKey.length() <= 0) {
              System.out.println("Please select a recipient key.");
              break;
            }
            pgp.getRecipientKeys().add(new Key(keyringDir, recipientKey));
            pgp.encrypt();
            if (pgp.getOutputFile().length() <= 0) {
              System.out.println("\nOutput Message:\n");
              System.out.println(new String(pgp.getOutputMessage()));
            }
          } else if (command == 2) { //Decrypt
            if (privateKey.length() <= 0) {
              System.out.println("Please select a private key.");
              break;
            }
            pgp.getKeys().add(new Key(keyringDir, privateKey));
            pgp.getKeys().item(0).setPassphrase(passphrase);
            pgp.setASCIIArmor(true);
            pgp.decrypt();
            if (pgp.getOutputFile().length() <= 0) {
              System.out.println("\nOutput Message:\n");
              System.out.println(new String(pgp.getOutputMessage()));
            }
          } else if (command == 3) { //Sign
            if (privateKey.length() <= 0) {
              System.out.println("Please select a private key.");
              break;
            }
            pgp.getKeys().add(new Key(keyringDir, privateKey));
            pgp.getKeys().item(0).setPassphrase(passphrase);
            pgp.sign();
            if (pgp.getOutputFile().length() <= 0) {
              System.out.println("\nOutput Message:\n");
              System.out.println(new String(pgp.getOutputMessage()));
            }
          } else if (command == 4) { //Verify
            if (signerKey.length() <= 0) {
              System.out.println("Please select a signer key.");
              break;
            }
            pgp.getSignerKeys().add(new Key(keyringDir, signerKey));
            pgp.verifySignature();
            if (pgp.getOutputFile().length() <= 0) {
              System.out.println("\nOutput Message:\n");
              System.out.println(new String(pgp.getOutputMessage()));
            }
          } else if (command == 5) { //Sign and Encrypt
            if (privateKey.length() <= 0) {
              System.out.println("Please select a private key.");
              break;
            }
            if (recipientKey.length() <= 0) {
              System.out.println("Please select a recipient key.");
              break;
            }
            pgp.getKeys().add(new Key(keyringDir, privateKey));
            pgp.getKeys().item(0).setPassphrase(passphrase);
            pgp.getRecipientKeys().add(new Key(keyringDir, recipientKey));
            pgp.signAndEncrypt();
            if (pgp.getOutputFile().length() <= 0) {
              System.out.println("\nOutput Message:\n");
              System.out.println(new String(pgp.getOutputMessage()));
            }
          } else if (command == 6) { //Decrypt and Verify
            if (privateKey.length() <= 0) {
              System.out.println("Please select a private key.");
              break;
            }
            if (signerKey.length() <= 0) {
              System.out.println("Please select a signer key.");
              break;
            }
            pgp.getKeys().add(new Key(keyringDir, privateKey));
            pgp.getKeys().item(0).setPassphrase(passphrase);
            pgp.getSignerKeys().add(new Key(keyringDir, signerKey));
            pgp.decryptAndVerifySignature();
            if (pgp.getOutputFile().length() <= 0) {
              System.out.println("\nOutput Message:\n");
              System.out.println(new String(pgp.getOutputMessage()));
            }        
          }
        } catch(Exception ex) {
          System.out.println("Error: " + ex.getMessage());
        }
      }

    } catch(IPWorksOpenPGPException ex) {
      displayError(ex);
    } catch(Exception ex) {
      displayError(ex);
    }
  }

  public static void main(String[] args) {
    new openpgp();
  }
}


class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksOpenPGPException) {
      System.out.print(" (" + ((IPWorksOpenPGPException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: IPWorks OpenPGP Samples/OpenPGP/c++/openpgp.cpp
/*
 * IPWorks OpenPGP 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks OpenPGP in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksopenpgp
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include "../../include/ipworksopenpgp.h"
#define LINE_LEN 100
#define MESSAGE_LEN 1024

class MyKeyMgr : public KeyMgr 
{
public:
  virtual int FireKeyList(KeyMgrKeyListEventParams *e) 
  {
    printf("%-58s %-8s %-12s\n", e->UserId, e->KeyId, e->HasSecretKey ? "private" : "");  
    return 0;
  }
};

int main(int argc, char **argv)
{
  OpenPGP pgp;
  MyKeyMgr keymgr;

  int ret_code = 0;
  char command[LINE_LEN];     // user's command
  char keyringDir[LINE_LEN];
  char passphrase[LINE_LEN];
  char privateKey[LINE_LEN];
  char recipientKey[LINE_LEN];
  char signerKey[LINE_LEN];

  printf("Keyring Directory: ");
  fgets(keyringDir, LINE_LEN, stdin);
  keyringDir[strlen(keyringDir)-1] = '\0';
  keymgr.LoadKeyring(keyringDir);

  printf("Create a new key? (y/n) ");
  fgets(command, LINE_LEN, stdin);
  command[strlen(command)-1] = '\0';
  if (tolower(command[0]) == 'y') 
  {
    while (1) 
    {
      printf("\nListing Keys...\n");
      printf("%-58s %-8s %-12s\n", "User Id", "Key Id", "Private Key"); 
      keymgr.ListKeys();
      printf("\nPlease enter a command: (c: create key, d: delete key, s: save keyring, q: cancel) ");
      fgets(command, LINE_LEN, stdin);
      command[strlen(command)-1] = '\0';
      if (tolower(command[0]) == 'c') 
      {
        char userid[LINE_LEN];
        printf("User Id: ");
        fgets(userid, LINE_LEN, stdin);
        userid[strlen(userid)-1] = '\0';
        printf("Passphrase: ");
        fgets(passphrase, LINE_LEN, stdin);
        passphrase[strlen(passphrase)-1] = '\0';
        ret_code = keymgr.CreateKey(userid, passphrase);
        if (ret_code) goto done;
      } 
      else if (tolower(command[0]) == 'd') 
      {
        char userid[LINE_LEN];
        printf("User Id: ");
        fgets(userid, LINE_LEN, stdin);
        userid[strlen(userid)-1] = '\0';
        ret_code = keymgr.DeleteKey(userid);
        if (ret_code) goto done;
      }
      else if (tolower(command[0]) == 's') 
      {
        ret_code = keymgr.SaveKeyring(keyringDir);
        if (ret_code) goto done;
        break;
      }
      else if (tolower(command[0]) == 'q') 
      {
        break;
      }
      else 
      {
        printf("Invalid command.");
      }
    }
  }
  
  printf("\nListing Keys...\n");
  printf("%-58s %-8s %-12s\n", "User Id", "Key Id", "Private Key"); 
  keymgr.ListKeys();

  printf("\nPrivate Key User Id (Used to sign outgoing messages and decrypt incoming messages): ");
  fgets(privateKey, LINE_LEN, stdin);
  privateKey[strlen(privateKey)-1] = '\0';
  printf("Passphrase: ");
  fgets(passphrase, LINE_LEN, stdin);
  passphrase[strlen(passphrase)-1] = '\0';
  printf("Recipient Key User Id (Used to encrypt outgoing messages): ");
  fgets(recipientKey, LINE_LEN, stdin);
  recipientKey[strlen(recipientKey)-1] = '\0';
  printf("Signer Key User Id (Used to verify the signature of an incoming message): ");
  fgets(signerKey, LINE_LEN, stdin);
  signerKey[strlen(signerKey)-1] = '\0';

  while (1) 
  {
    printf("Please enter a command: (0: quit, 1: encrypt, 2: decrypt, 3: sign, 4: verify, 5: sign and encrypt, 6: decrypt and verify) ");
    fgets(command, LINE_LEN, stdin);
    command[strlen(command)-1] = '\0';
    int cmd = atoi(command);

    if (cmd < 0 || cmd > 6) 
    {
      printf("Invalid Command.\n");
      continue;
    }
    if (cmd == 0) 
      break;

    pgp.Reset();
    
    printf("Read from file or string? (f/s) ");
    fgets(command, LINE_LEN, stdin);
    command[strlen(command)-1] = '\0';
    if (tolower(command[0]) == 'f') 
    {
      printf("Input File: ");
      fgets(command, LINE_LEN, stdin);
      command[strlen(command)-1] = '\0';
      pgp.SetInputFile(command);

      printf("Output File: ");
        fgets(command, LINE_LEN, stdin);
        command[strlen(command)-1] = '\0';
      pgp.SetOutputFile(command);
            
      printf("Overwrite file? (y/n) ");
        fgets(command, LINE_LEN, stdin);
        command[strlen(command)-1] = '\0';
      pgp.SetOverwrite(tolower(command[0]) == 'y');
    }
    else 
    {
      printf("Please enter the message. When finished enter \":q\" on a line by itself:\n");
      char message[MESSAGE_LEN];
        message[0] = '\0';
        while(fgets(command, LINE_LEN, stdin))
        {
            command[strlen(command)-1] = '\0';
            if (strcmp(command, ":q") == 0)
                break;
            strcat(message, command);
            strcat(message, "\n");
        }
      pgp.SetInputMessage(message, strlen(message));
    }

    if (cmd == 1) //Encrypt
    {
      if (strlen(recipientKey) <= 0) 
      {
        printf("Please select a recipient key.\n");
        break;
      }
      pgp.SetASCIIArmor(true);
      pgp.SetRecipientKeyCount(1);
      pgp.SetRecipientKeyKeyring(0, keyringDir);
      pgp.SetRecipientKeyUserId(0, recipientKey);
      ret_code = pgp.Encrypt();
      if (ret_code) goto done;

      if (strlen(pgp.GetOutputFile()) <= 0) 
      {
        printf("\nEncrypted Message:\n");
        char* output;
        int len;
        pgp.GetOutputMessage(output, len);
        printf("%s\n", output);
      }
    }
    else if (cmd == 2) //Decrypt
    {
      if (strlen(privateKey) <= 0) 
      {
        printf("Please select a private key.\n");
        break;
      }

      pgp.SetKeyCount(1);
      pgp.SetKeyKeyring(0, keyringDir);
      pgp.SetKeyUserId(0, privateKey);
      pgp.SetKeyPassphrase(0, passphrase);
      ret_code = pgp.Decrypt();
      if (ret_code) goto done;
      
      if (strlen(pgp.GetOutputFile()) <= 0) 
      {
        printf("\nDecrypted Message:\n");
        char* output;
        int len;
        pgp.GetOutputMessage(output, len);
        printf("%s\n", output);
      }
    }
    else if (cmd == 3) //Sign
    {
      if (strlen(privateKey) <= 0) 
      {
        printf("Please select a private key.\n");
        break;
      }

      pgp.SetASCIIArmor(true);
      pgp.SetKeyCount(1);
      pgp.SetKeyKeyring(0, keyringDir);
      pgp.SetKeyUserId(0, privateKey);
      pgp.SetKeyPassphrase(0, passphrase);
      ret_code = pgp.Sign();
      if (ret_code) goto done;
      
      if (strlen(pgp.GetOutputFile()) <= 0) 
      {
        printf("\nSigned Message:\n");
        char* output;
        int len;
        pgp.GetOutputMessage(output, len);
        printf("%s\n", output);
      }
    }
    else if (cmd == 4) //Verify
    {
      if (strlen(signerKey) <= 0) 
      {
        printf("Please select a signer key.\n");
        break;
      }

      pgp.SetSignerKeyCount(1);
      pgp.SetSignerKeyKeyring(0, keyringDir);
      pgp.SetSignerKeyUserId(0, signerKey);
      ret_code = pgp.VerifySignature();
      if (ret_code) goto done;
      
      if (strlen(pgp.GetOutputFile()) <= 0) 
      {
        printf("\nVerified Message:\n");
        char* output;
        int len;
        pgp.GetOutputMessage(output, len);
        printf("%s\n", output);
      }
    }
    else if (cmd == 5) //Sign and Encrypt
    {
      if (strlen(privateKey) <= 0) 
      {
        printf("Please select a private key.\n");
        break;
      }
      if (strlen(recipientKey) <= 0) 
      {
        printf("Please select a recipient key.\n");
        break;
      }
      if (strlen(signerKey) <= 0) 
      {
        printf("Please select a signer key.\n");
        break;
      }

      pgp.SetASCIIArmor(true);
      pgp.SetKeyCount(1);
      pgp.SetKeyKeyring(0, keyringDir);
      pgp.SetKeyUserId(0, privateKey);
      pgp.SetKeyPassphrase(0, passphrase);
      pgp.SetRecipientKeyCount(1);
      pgp.SetRecipientKeyKeyring(0, keyringDir);
      pgp.SetRecipientKeyUserId(0, recipientKey);
      ret_code = pgp.SignAndEncrypt();
      if (ret_code) goto done;
      
      if (strlen(pgp.GetOutputFile()) <= 0) 
      {
        printf("\nSigned & Encrypted Message:\n");
        char* output;
        int len;
        pgp.GetOutputMessage(output, len);
        printf("%s\n", output);
      }
    }
    else if (cmd == 6) //Decrypt and Verify
    {
      if (strlen(privateKey) <= 0) 
      {
        printf("Please select a private key.\n");
        break;
      }

      pgp.SetKeyCount(1);
      pgp.SetKeyKeyring(0, keyringDir);
      pgp.SetKeyUserId(0, privateKey);
      pgp.SetKeyPassphrase(0, passphrase);       
      pgp.SetSignerKeyCount(1);
      pgp.SetSignerKeyKeyring(0, keyringDir);
      pgp.SetSignerKeyUserId(0, signerKey);
      ret_code = pgp.DecryptAndVerifySignature();
      if (ret_code) goto done;
      
      if (strlen(pgp.GetOutputFile()) <= 0) 
      {
        printf("\nDecrypted & Verified Message:\n");
        char* output;
        int len;
        pgp.GetOutputMessage(output, len);
        printf("%s\n", output);
      }
    }
  }

done:
    if (ret_code)     // Got an error.  The user is done.
    {
        printf( "\nError: %d", ret_code );
        if (pgp.GetLastError())
        {
            printf( " \"%s\"\n", pgp.GetLastError() );
        }
    }

    fprintf(stderr, "\npress <return> to continue...\n");
    getchar();
    exit(ret_code);
    return 0;
}



